bee-ledger/src/workers/consensus/white_flag.rs

```diff
@@ -68,70 +58,155 @@ fn apply_regular_essence<B: StorageBackend>(
                 }
             }
-            Input::Treasury(_) => {
+            _ => {
                 return Err(Error::UnsupportedInputKind(input.kind()));
             }
         };
 
-        let _essence_hash = TransactionEssence::from(essence.clone()).hash();
+        consumed_outputs.push((*output_id, consumed_output));
+    }
 
-        match consumed_output.inner() {
-            #[cfg(feature = "cpt2")]
-            Output::SignatureLockedSingle(_) => {
-                // TODO
-            }
-            #[cfg(feature = "cpt2")]
-            Output::SignatureLockedDustAllowance(_) => {
-                // TODO
-            }
-            Output::Treasury(_) => return Err(Error::UnsupportedOutputKind(consumed_output.inner().kind())),
-            Output::Basic(_) => {
-                // TODO
-            }
-            Output::Alias(_) => {
-                // TODO
-            }
-            Output::Foundry(_) => {
-                // TODO
+    let inputs: Vec<(OutputId, &Output)> = consumed_outputs
+        .iter()
+        .map(|(output_id, created_output)| (*output_id, created_output.inner()))
+        .collect();
+
+    let mut context = ValidationContext::new(
+        transaction_id,
+        essence,
+        inputs.iter().map(|(output_id, input)| (output_id, *input)),
+        unlock_blocks,
+        metadata.milestone_index,
+        metadata.milestone_timestamp,
+    );
+
+    // Validation of the inputs commitment.
+    if context.essence.inputs_commitment() != &context.inputs_commitment {
+        return Ok(ConflictReason::InputsCommitmentsMismatch);
+    }
+
+    // Validation of inputs.
+    for ((output_id, consumed_output), unlock_block) in inputs.iter().zip(unlock_blocks.iter()) {
+        let (conflict, amount, consumed_native_tokens, unlock_conditions) = match consumed_output {
+            Output::Basic(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Alias(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Foundry(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Nft(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            _ => return Err(Error::UnsupportedOutputKind(consumed_output.kind())),
+        };
```

Comment:
Maybe some shared method here could help with this repetition? I'm not sure whether its worth the effort or the indirection, but this code feels a little like it could do with some clean-up.

Reply:
We actually have these methods on the `Output` type itself. But they internally match so it would mean matching 4 times for each output...

Reply:
Ah, that's a pain!